File: archive/typescript/src/core/profiles.ts
import { homedir } from "node:os";
import { dirname, join } from "node:path";
import { mkdir, readFile, writeFile } from "node:fs/promises";

export type AgentboxProfile = {
  name: string;
  baseUrl: string;
  apiKey: string;
};

export type ProfileStore = {
  activeProfileName: string | null;
  profiles: AgentboxProfile[];
};

export type ResolvedProfile = AgentboxProfile & {
  source: "config" | "env" | "legacy-env";
};

type RawProfileRecord = {
  base_url?: unknown;
  baseUrl?: unknown;
  api_key?: unknown;
  apiKey?: unknown;
};

type ProfileStoreFile = {
  active_profile?: unknown;
  current_profile?: unknown;
  profiles?: unknown;
};

function normalizeProfile(name: string, value: RawProfileRecord): AgentboxProfile | null {
  const trimmedName = name.trim();
  const baseUrl = typeof value.base_url === "string"
    ? value.base_url
    : typeof value.baseUrl === "string"
      ? value.baseUrl
      : null;
  const apiKey = typeof value.api_key === "string"
    ? value.api_key
    : typeof value.apiKey === "string"
      ? value.apiKey
      : null;

  if (!trimmedName || !baseUrl || !apiKey) return null;

  return {
    name: trimmedName,
    baseUrl: baseUrl.trim().replace(/\/+$/, ""),
    apiKey: apiKey.trim()
  };
}

function parseProfilesRecord(value: unknown): AgentboxProfile[] {
  if (Array.isArray(value)) {
    return value.flatMap((entry) => {
      if (!entry || typeof entry !== "object") return [];
      const candidate = entry as { name?: unknown } & RawProfileRecord;
      if (typeof candidate.name !== "string") return [];
      const profile = normalizeProfile(candidate.name, candidate);
      return profile ? [profile] : [];
    });
  }

  if (!value || typeof value !== "object") return [];

  return Object.entries(value as Record<string, RawProfileRecord>).flatMap(([name, record]) => {
    if (!record || typeof record !== "object") return [];
    const profile = normalizeProfile(name, record);
    return profile ? [profile] : [];
  });
}

function dedupeProfiles(profiles: AgentboxProfile[]): AgentboxProfile[] {
  const byName = new Map<string, AgentboxProfile>();
  for (const profile of profiles) {
    byName.set(profile.name, profile);
  }
  return [...byName.values()].sort((a, b) => a.name.localeCompare(b.name));
}

export function parseProfilesConfig(raw: string | undefined): AgentboxProfile[] {
  if (!raw) return [];

  const trimmed = raw.trim();
  if (!trimmed) return [];

  const parsed = JSON.parse(trimmed) as unknown;
  return dedupeProfiles(parseProfilesRecord(parsed));
}

export function defaultConfigDir(): string {
  if (process.env.AGENTBOX_CONFIG_DIR?.trim()) return process.env.AGENTBOX_CONFIG_DIR.trim();
  if (process.platform === "win32" && process.env.APPDATA?.trim()) {
    return join(process.env.APPDATA.trim(), "agentbox");
  }
  if (process.platform === "darwin") {
    return join(homedir(), "Library", "Application Support", "agentbox");
  }
  if (process.env.XDG_CONFIG_HOME?.trim()) {
    return join(process.env.XDG_CONFIG_HOME.trim(), "agentbox");
  }
  return join(homedir(), ".config", "agentbox");
}

export function defaultConfigPath(): string {
  return join(defaultConfigDir(), "profiles.json");
}

export async function readProfileStore(): Promise<ProfileStore> {
  try {
    const raw = await readFile(defaultConfigPath(), "utf8");
    const parsed = JSON.parse(raw) as ProfileStoreFile;
    const profiles = dedupeProfiles(parseProfilesRecord(parsed.profiles));
    const candidate = typeof parsed.active_profile === "string"
      ? parsed.active_profile
      : typeof parsed.current_profile === "string"
        ? parsed.current_profile
        : null;
    const activeProfileName = candidate && profiles.some((profile) => profile.name === candidate)
      ? candidate
      : null;
    return { activeProfileName, profiles };
  } catch (error) {
    const code = (error as NodeJS.ErrnoException | undefined)?.code;
    if (code === "ENOENT") return { activeProfileName: null, profiles: [] };
    throw error;
  }
}

async function writeProfileStore(store: ProfileStore): Promise<void> {
  const filePath = defaultConfigPath();
  await mkdir(dirname(filePath), { recursive: true });
  const normalizedProfiles = dedupeProfiles(store.profiles);
  const activeProfileName = store.activeProfileName && normalizedProfiles.some((profile) => profile.name === store.activeProfileName)
    ? store.activeProfileName
    : normalizedProfiles[0]?.name ?? null;
  const payload = {
    active_profile: activeProfileName,
    profiles: Object.fromEntries(normalizedProfiles.map((profile) => [
      profile.name,
      {
        base_url: profile.baseUrl,
        api_key: profile.apiKey
      }
    ]))
  };
  await writeFile(filePath, `${JSON.stringify(payload, null, 2)}\n`, { mode: 0o600 });
}

export async function saveProfile(profile: AgentboxProfile, options?: { activate?: boolean }): Promise<ProfileStore> {
  const store = await readProfileStore();
  const profiles = [
    ...store.profiles.filter((entry) => entry.name !== profile.name),
    normalizeProfile(profile.name, { base_url: profile.baseUrl, api_key: profile.apiKey }) ?? profile
  ];
  const nextStore: ProfileStore = {
    activeProfileName: options?.activate ? profile.name : store.activeProfileName,
    profiles
  };
  await writeProfileStore(nextStore);
  return readProfileStore();
}

export async function removeProfile(name: string): Promise<ProfileStore> {
  const store = await readProfileStore();
  const profiles = store.profiles.filter((profile) => profile.name !== name);
  if (profiles.length === store.profiles.length) {
    throw new Error(`Unknown Agentbox profile "${name}".`);
  }
  const nextStore: ProfileStore = {
    activeProfileName: store.activeProfileName === name ? (profiles[0]?.name ?? null) : store.activeProfileName,
    profiles
  };
  await writeProfileStore(nextStore);
  return readProfileStore();
}

export async function setActiveProfile(name: string): Promise<ProfileStore> {
  const store = await readProfileStore();
  if (!store.profiles.some((profile) => profile.name === name)) {
    throw new Error(`Unknown Agentbox profile "${name}".`);
  }
  await writeProfileStore({ ...store, activeProfileName: name });
  return readProfileStore();
}

export function maskSecret(secret: string, options?: { visiblePrefix?: number; visibleSuffix?: number }): string {
  const visiblePrefix = options?.visiblePrefix ?? 3;
  const visibleSuffix = options?.visibleSuffix ?? 2;
  if (secret.length <= visiblePrefix + visibleSuffix) return "*".repeat(Math.max(secret.length, 4));
  return `${secret.slice(0, visiblePrefix)}${"*".repeat(secret.length - visiblePrefix - visibleSuffix)}${secret.slice(-visibleSuffix)}`;
}

export function sanitizeUrl(url: URL): string {
  const clone = new URL(url.toString());
  const key = clone.searchParams.get("key");
  if (key) clone.searchParams.set("key", maskSecret(key));
  return clone.toString();
}

export async function resolveProfileSelection(options?: { profileName?: string | null }): Promise<ResolvedProfile | null> {
  const envProfiles = parseProfilesConfig(process.env.AGENTBOX_PROFILES);
  const selectedName = options?.profileName ?? process.env.AGENTBOX_PROFILE ?? null;

  if (envProfiles.length > 0) {
    if (selectedName) {
      const match = envProfiles.find((profile) => profile.name === selectedName);
      if (!match) throw new Error(`Unknown Agentbox profile "${selectedName}".`);
      return { ...match, source: "env" };
    }
    return { ...envProfiles[0], source: "env" };
  }

  const store = await readProfileStore();
  if (store.profiles.length > 0) {
    const activeName = selectedName ?? store.activeProfileName ?? store.profiles[0]?.name ?? null;
    const match = activeName ? store.profiles.find((profile) => profile.name === activeName) : null;
    if (selectedName && !match) throw new Error(`Unknown Agentbox profile "${selectedName}".`);
    if (match) return { ...match, source: "config" };
  }

  const legacyBaseUrl = process.env.AGENTBOX_BASE_URL ?? process.env.AGENTBOX_URL;
  const legacyApiKey = process.env.AGENTBOX_API_KEY;
  if (!legacyBaseUrl || !legacyApiKey) return null;
  if (selectedName && selectedName !== "default") {
    throw new Error(`Unknown Agentbox profile "${selectedName}".`);
  }

  return {
    name: selectedName ?? "default",
    baseUrl: legacyBaseUrl.trim().replace(/\/+$/, ""),
    apiKey: legacyApiKey.trim(),
    source: "legacy-env"
  };
}

File: internal/agentbox/profiles/profiles.go
package profiles

import (
	"encoding/json"
	"errors"
	"fmt"
	"net/url"
	"os"
	"path/filepath"
	"runtime"
	"sort"
	"strings"
)

type Profile struct {
	Name       string `json:"name"`
	BaseURL    string `json:"base_url"`
	APIKey     string `json:"api_key"`
	TenantID   string `json:"tenant_id,omitempty"`
	TenantSlug string `json:"tenant_slug,omitempty"`
	TenantName string `json:"tenant_name,omitempty"`
	UserID     string `json:"user_id,omitempty"`
	KeyName    string `json:"key_name,omitempty"`
	AuthType   string `json:"auth_type,omitempty"`
}

type Store struct {
	ActiveProfileName string
	Profiles          []Profile
}

type ResolvedProfile struct {
	Profile
	Source string `json:"source"`
}

func DefaultConfigDir() string {
	if value := strings.TrimSpace(os.Getenv("AGENTBOX_CONFIG_DIR")); value != "" {
		return value
	}
	if runtime.GOOS == "windows" {
		if appData := strings.TrimSpace(os.Getenv("APPDATA")); appData != "" {
			return filepath.Join(appData, "agentbox")
		}
	}
	home, _ := os.UserHomeDir()
	if runtime.GOOS == "darwin" {
		return filepath.Join(home, "Library", "Application Support", "agentbox")
	}
	if xdg := strings.TrimSpace(os.Getenv("XDG_CONFIG_HOME")); xdg != "" {
		return filepath.Join(xdg, "agentbox")
	}
	return filepath.Join(home, ".config", "agentbox")
}

func DefaultConfigPath() string {
	return filepath.Join(DefaultConfigDir(), "profiles.json")
}

func ParseProfilesConfig(raw string) ([]Profile, error) {
	if strings.TrimSpace(raw) == "" {
		return nil, nil
	}
	var value any
	if err := json.Unmarshal([]byte(strings.TrimSpace(raw)), &value); err != nil {
		return nil, err
	}
	return dedupeProfiles(parseProfilesRecord(value)), nil
}

func ReadStore() (Store, error) {
	bytes, err := os.ReadFile(DefaultConfigPath())
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return Store{}, nil
		}
		return Store{}, err
	}
	var raw map[string]any
	if err := json.Unmarshal(bytes, &raw); err != nil {
		return Store{}, err
	}
	profiles := dedupeProfiles(parseProfilesRecord(raw["profiles"]))
	active := stringField(raw, "active_profile", "current_profile")
	if active != "" && !hasProfile(profiles, active) {
		active = ""
	}
	return Store{ActiveProfileName: active, Profiles: profiles}, nil
}

func SaveProfile(profile Profile, activate bool) (Store, error) {
	store, err := ReadStore()
	if err != nil {
		return Store{}, err
	}
	normalized, ok := normalizeProfile(profile.Name, map[string]any{
		"base_url":    profile.BaseURL,
		"api_key":     profile.APIKey,
		"tenant_id":   profile.TenantID,
		"tenant_slug": profile.TenantSlug,
		"tenant_name": profile.TenantName,
		"user_id":     profile.UserID,
		"key_name":    profile.KeyName,
		"auth_type":   profile.AuthType,
	})
	if !ok {
		normalized = profile
	}
	next := make([]Profile, 0, len(store.Profiles)+1)
	for _, existing := range store.Profiles {
		if existing.Name != normalized.Name {
			next = append(next, existing)
		}
	}
	next = append(next, normalized)
	active := store.ActiveProfileName
	if activate {
		active = normalized.Name
	}
	if err := writeStore(Store{ActiveProfileName: active, Profiles: next}); err != nil {
		return Store{}, err
	}
	return ReadStore()
}

func RemoveProfile(name string) (Store, error) {
	store, err := ReadStore()
	if err != nil {
		return Store{}, err
	}
	next := make([]Profile, 0, len(store.Profiles))
	for _, profile := range store.Profiles {
		if profile.Name != name {
			next = append(next, profile)
		}
	}
	if len(next) == len(store.Profiles) {
		return Store{}, fmt.Errorf("Unknown Agentbox profile %q.", name)
	}
	active := store.ActiveProfileName
	if active == name {
		active = ""
		if len(next) > 0 {
			active = next[0].Name
		}
	}
	if err := writeStore(Store{ActiveProfileName: active, Profiles: next}); err != nil {
		return Store{}, err
	}
	return ReadStore()
}

func SetActiveProfile(name string) (Store, error) {
	store, err := ReadStore()
	if err != nil {
		return Store{}, err
	}
	if !hasProfile(store.Profiles, name) {
		return Store{}, fmt.Errorf("Unknown Agentbox profile %q.", name)
	}
	if err := writeStore(Store{ActiveProfileName: name, Profiles: store.Profiles}); err != nil {
		return Store{}, err
	}
	return ReadStore()
}

func Resolve(selection string) (*ResolvedProfile, error) {
	envProfiles, err := ParseProfilesConfig(os.Getenv("AGENTBOX_PROFILES"))
	if err != nil {
		return nil, err
	}
	selected := selection
	if selected == "" {
		selected = os.Getenv("AGENTBOX_PROFILE")
	}
	if len(envProfiles) > 0 {
		if selected != "" {
			match := findProfile(envProfiles, selected)
			if match == nil {
				return nil, fmt.Errorf("Unknown Agentbox profile %q.", selected)
			}
			return &ResolvedProfile{Profile: *match, Source: "env"}, nil
		}
		return &ResolvedProfile{Profile: envProfiles[0], Source: "env"}, nil
	}
	store, err := ReadStore()
	if err != nil {
		return nil, err
	}
	if len(store.Profiles) > 0 {
		active := selected
		if active == "" {
			active = store.ActiveProfileName
		}
		if active == "" {
			active = store.Profiles[0].Name
		}
		match := findProfile(store.Profiles, active)
		if selected != "" && match == nil {
			return nil, fmt.Errorf("Unknown Agentbox profile %q.", selected)
		}
		if match != nil {
			return &ResolvedProfile{Profile: *match, Source: "config"}, nil
		}
	}
	legacyBaseURL := os.Getenv("AGENTBOX_BASE_URL")
	if legacyBaseURL == "" {
		legacyBaseURL = os.Getenv("AGENTBOX_URL")
	}
	legacyAPIKey := os.Getenv("AGENTBOX_API_KEY")
	if legacyBaseURL == "" || legacyAPIKey == "" {
		return nil, nil
	}
	if selected != "" && selected != "default" {
		return nil, fmt.Errorf("Unknown Agentbox profile %q.", selected)
	}
	return &ResolvedProfile{
		Profile: Profile{Name: defaultString(selected, "default"), BaseURL: trimURL(legacyBaseURL), APIKey: strings.TrimSpace(legacyAPIKey)},
		Source:  "legacy-env",
	}, nil
}

func MaskSecret(secret string) string {
	return MaskSecretWith(secret, 3, 2)
}

func MaskSecretWith(secret string, prefix int, suffix int) string {
	if len(secret) <= prefix+suffix {
		if len(secret) > 4 {
			return strings.Repeat("*", len(secret))
		}
		return "****"
	}
	return secret[:prefix] + strings.Repeat("*", len(secret)-prefix-suffix) + secret[len(secret)-suffix:]
}

func SanitizeURL(value string) string {
	parsed, err := url.Parse(value)
	if err != nil {
		return value
	}
	key := parsed.Query().Get("key")
	if key == "" {
		return parsed.String()
	}
	query := parsed.Query()
	query.Set("key", MaskSecret(key))
	parsed.RawQuery = query.Encode()
	return parsed.String()
}

func writeStore(store Store) error {
	profiles := dedupeProfiles(store.Profiles)
	active := store.ActiveProfileName
	if active != "" && !hasProfile(profiles, active) {
		active = ""
	}
	if active == "" && len(profiles) > 0 {
		active = profiles[0].Name
	}
	var activeProfile *string
	if active != "" {
		activeProfile = &active
	}
	payload := struct {
		ActiveProfile *string                   `json:"active_profile"`
		Profiles      map[string]map[string]any `json:"profiles"`
	}{
		ActiveProfile: activeProfile,
		Profiles:      map[string]map[string]any{},
	}
	for _, profile := range profiles {
		record := map[string]any{
			"base_url": profile.BaseURL,
			"api_key":  profile.APIKey,
		}
		setOptionalField(record, "tenant_id", profile.TenantID)
		setOptionalField(record, "tenant_slug", profile.TenantSlug)
		setOptionalField(record, "tenant_name", profile.TenantName)
		setOptionalField(record, "user_id", profile.UserID)
		setOptionalField(record, "key_name", profile.KeyName)
		setOptionalField(record, "auth_type", profile.AuthType)
		payload.Profiles[profile.Name] = record
	}
	bytes, err := json.MarshalIndent(payload, "", "  ")
	if err != nil {
		return err
	}
	path := DefaultConfigPath()
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return err
	}
	return os.WriteFile(path, append(bytes, '\n'), 0o600)
}

func parseProfilesRecord(value any) []Profile {
	switch typed := value.(type) {
	case []any:
		var profiles []Profile
		for _, entry := range typed {
			record, ok := entry.(map[string]any)
			if !ok {
				continue
			}
			name, _ := record["name"].(string)
			if profile, ok := normalizeProfile(name, record); ok {
				profiles = append(profiles, profile)
			}
		}
		return profiles
	case map[string]any:
		var profiles []Profile
		for name, entry := range typed {
			record, ok := entry.(map[string]any)
			if !ok {
				continue
			}
			if profile, ok := normalizeProfile(name, record); ok {
				profiles = append(profiles, profile)
			}
		}
		return profiles
	default:
		return nil
	}
}

func normalizeProfile(name string, record map[string]any) (Profile, bool) {
	name = strings.TrimSpace(name)
	baseURL := stringField(record, "base_url", "baseUrl")
	apiKey := stringField(record, "api_key", "apiKey")
	if name == "" || baseURL == "" || apiKey == "" {
		return Profile{}, false
	}
	return Profile{
		Name:       name,
		BaseURL:    trimURL(baseURL),
		APIKey:     strings.TrimSpace(apiKey),
		TenantID:   strings.TrimSpace(stringField(record, "tenant_id", "tenantId")),
		TenantSlug: strings.TrimSpace(stringField(record, "tenant_slug", "tenantSlug")),
		TenantName: strings.TrimSpace(stringField(record, "tenant_name", "tenantName")),
		UserID:     strings.TrimSpace(stringField(record, "user_id", "userId")),
		KeyName:    strings.TrimSpace(stringField(record, "key_name", "keyName")),
		AuthType:   strings.TrimSpace(stringField(record, "auth_type", "authType")),
	}, true
}

func setOptionalField(record map[string]any, name string, value string) {
	value = strings.TrimSpace(value)
	if value != "" {
		record[name] = value
	}
}

func dedupeProfiles(profiles []Profile) []Profile {
	byName := map[string]Profile{}
	for _, profile := range profiles {
		byName[profile.Name] = profile
	}
	names := make([]string, 0, len(byName))
	for name := range byName {
		names = append(names, name)
	}
	sort.Strings(names)
	result := make([]Profile, 0, len(names))
	for _, name := range names {
		result = append(result, byName[name])
	}
	return result
}

func stringField(record map[string]any, names ...string) string {
	for _, name := range names {
		if value, ok := record[name].(string); ok {
			return value
		}
	}
	return ""
}

func hasProfile(profiles []Profile, name string) bool {
	return findProfile(profiles, name) != nil
}

func findProfile(profiles []Profile, name string) *Profile {
	for i := range profiles {
		if profiles[i].Name == name {
			return &profiles[i]
		}
	}
	return nil
}

func trimURL(value string) string {
	return strings.TrimRight(strings.TrimSpace(value), "/")
}

func defaultString(value string, fallback string) string {
	if value == "" {
		return fallback
	}
	return value
}
